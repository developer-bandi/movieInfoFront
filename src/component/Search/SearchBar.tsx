import { useState } from 'react';
import styled from 'styled-components';
import { css } from 'styled-components';
import { BsXLg } from 'react-icons/bs';
import { AiOutlineSearch } from 'react-icons/ai';

interface SearchBarProps {
  latest: string[];
  searchMovie: any;
  deleteComment: Function;
  settingvalue: React.ChangeEventHandler<HTMLInputElement>;
  value: string;
}

const SearchBar = ({
  latest,
  searchMovie,
  deleteComment,
  settingvalue,
  value,
}: SearchBarProps) => {
  const [focus, setFocus] = useState(false);
  const showLatest = () => {
    setFocus(!focus);
  };

  return (
    <SearchBarBlock>
      <InputBlock>
        <SearchInputBox
          type="text"
          onKeyPress={searchMovie}
          onChange={settingvalue}
          onClick={showLatest}
          value={value}
        ></SearchInputBox>
        <SearchButton onClick={searchMovie}></SearchButton>
      </InputBlock>
      <LatestSearchBlock active={focus}>
        {latest[0] === 'default' ? (
          <LatestNullSearchContent>
            {'최근 검색 내용이 없습니다'}
          </LatestNullSearchContent>
        ) : (
          latest.map((data, index) => (
            <LatestSearchContent key={index}>
              <div onClick={searchMovie}>{data}</div>
              <DeleteButton
                onClick={(e) => {
                  deleteComment(index);
                }}
              />
            </LatestSearchContent>
          ))
        )}
      </LatestSearchBlock>
    </SearchBarBlock>
  );
};

const SearchBarBlock = styled.div`
  background: ${(props) => props.theme.background};
  display: flex;
  justify-content: center;
  flex-direction: column;
  align-items: center;
  flex-wrap: wrap;
  padding-bottom: 5rem;
  padding-top: 5rem;
  border-bottom: 1px solid #e8e8e8;

  @media screen and (max-width: 500px) {
    padding-bottom: 1rem;
    padding-top: 1rem;
  }
`;
const InputBlock = styled.div`
  display: flex;
  border: 1px solid black;
  width: 55rem;
  background: white;
  @media screen and (max-width: 500px) {
    width: 20rem;
  }
`;

const SearchInputBox = styled.input`
  width: 50rem;
  font-size: 30px;
  height: 4rem;
  border: none;
  &:focus {
    outline: none;
  }

  @media screen and (max-width: 500px) {
    width: 17rem;
    height: 2rem;
    font-size: 20px;
  }
`;

const SearchButton = styled(AiOutlineSearch)`
  width: 3rem;
  height: 3rem;
  font-size: 30px;
  display: flex;
  justify-content: center;
  align-items: center;
  margin-top: 0.5rem;
  margin-left: 1rem;

  &:hover {
    color: #b2b2b2;
  }

  @media screen and (max-width: 500px) {
    width: 2rem;
    height: 2rem;
    font-size: 15px;
    margin-top: 0.1rem;
  }
`;

const LatestSearchBlock = styled.div<{ active: boolean }>`
  background: white;
  border: 1px solid black;
  border-top: none;
  display: none;
  width: 55rem;
  @media screen and (max-width: 500px) {
    width: 20rem;
    font-size: 15px;
  }
  ${(props) =>
    props.active &&
    css`
      display: block;
    `}
`;

const LatestSearchContent = styled.div`
  font-size: 15px;
  padding: 0.5rem;
  display: flex;
  align-items: center;
  & + & {
    border-top: 1px solid #e1e1e1;
  }
`;
const DeleteButton = styled(BsXLg)`
  width: 0.7rem;
  height: 0.7rem;
  margin-left: 0.5rem;
  m &:hover {
    color: gray;
  }
  @media screen and (max-width: 500px) {
    &:hover {
      color: black;
    }
  }
`;
const LatestNullSearchContent = styled.div`
  padding-top: 4rem;
  padding-bottom: 4rem;
  text-align: center;
  @media screen and (max-width: 500px) {
    padding-top: 2rem;
    padding-bottom: 2rem;
  }
`;

export default SearchBar;
