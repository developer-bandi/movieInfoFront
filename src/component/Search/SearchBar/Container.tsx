import { useState } from 'react';
import useSearchBar from './Hook';
import SearchBarPresentational from './Presentational';

const SearchBarContainer = () => {
  const { value, latest, searchMovies, settingvalue, deleteComment } =
    useSearchBar();
  const [focus, setFocus] = useState(false);
  return (
    <SearchBarPresentational
      latest={latest}
      searchMovies={searchMovies}
      deleteComment={deleteComment}
      settingvalue={settingvalue}
      value={value}
      focus={focus}
      setFocus={setFocus}
    />
  );
};

export default SearchBarContainer;
