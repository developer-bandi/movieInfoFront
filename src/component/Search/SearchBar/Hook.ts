import { useEffect, useState } from 'react';
import { useDispatch } from 'react-redux';
import { searchMovie } from '../../../store/movieSearch/Reducer';

const useSearchBar = () => {
  const [value, setvalue] = useState('');
  const [latest, setlatest] = useState<string[]>([]);
  const dispatch = useDispatch();

  useEffect(() => {
    const data = localStorage.getItem('latest');
    if (data !== null) {
      setlatest(JSON.parse(data));
    }
  }, []);
  const searchMovies = (e: {
    key?: string;
    type: string;
    target: { innerText: string; tagName: string };
  }) => {
    if (
      (e.type === 'click' && e.key === undefined) ||
      (e.type === 'keypress' && e.key === 'Enter')
    ) {
      //클릭 엔터 모두 영화를 검색할수 있게 합니다.
      let tempValue = value;
      if (e.target.tagName === 'DIV') {
        //이벤트 발생이 div태그 즉 페이지 버튼에 의한 것이면 인덱스를 따로 가져옵니다.
        tempValue = e.target.innerText;
      }
      dispatch(searchMovie({ keyword: tempValue, page: 1 }));

      if (latest.length === 5) {
        setlatest([...latest.slice(1, 5), tempValue]);
        localStorage.setItem(
          'latest',
          JSON.stringify([...latest.slice(1, 5), tempValue])
        );
      } else {
        setlatest([...latest, tempValue]);
        localStorage.setItem('latest', JSON.stringify([...latest, tempValue]));
      } //최근검색어 처리 로직입니다.

      setvalue(tempValue);
    }
  };

  const settingvalue = (e: React.ChangeEvent<HTMLInputElement>) => {
    setvalue(e.target.value);
  };
  //검색어 입력값을 반영하는 함수압니다.

  const deleteComment = (index: number) => {
    latest.splice(index, 1);
    setlatest([...latest]);
    localStorage.setItem('latest', JSON.stringify(latest));
  };
  //최근검색어를 삭제하는 함수압니다.

  //최근 검색어 데이터를 불러옵니다.

  return {
    latest,
    setvalue,
    searchMovies,
    settingvalue,
    deleteComment,
    value,
  };
};

export default useSearchBar;
