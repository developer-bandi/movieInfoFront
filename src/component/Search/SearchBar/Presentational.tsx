import styles from './Style';

interface SearchBarProps {
  latest: string[];
  searchMovies: any;
  deleteComment: (index: number) => void;
  settingvalue: React.ChangeEventHandler<HTMLInputElement>;
  value: string;
  focus: boolean;
  setFocus: React.Dispatch<React.SetStateAction<boolean>>;
}

const SearchBarPresentational = ({
  latest,
  searchMovies,
  deleteComment,
  settingvalue,
  value,
  focus,
  setFocus,
}: SearchBarProps) => {
  return (
    <styles.MainBlock>
      <styles.InputBlock>
        <styles.SearchInput
          type="text"
          onKeyPress={(e) => searchMovies(e)}
          onChange={(e) => settingvalue(e)}
          onClick={() => setFocus(!focus)}
          value={value}
          data-testid={'searchInput'}
        ></styles.SearchInput>
        <styles.SearchButton
          data-testid={'searchButton'}
          onClick={searchMovies}
        ></styles.SearchButton>
      </styles.InputBlock>
      <styles.LatestSearchListBlock>
        {focus ? (
          latest.length === 0 ? (
            <styles.LatestSearchNull>
              {'최근 검색 내용이 없습니다'}
            </styles.LatestSearchNull>
          ) : (
            latest.map((data, index) => (
              <styles.LatestSearch key={index}>
                <div data-testid={'data'} onClick={searchMovies}>
                  {data}
                </div>
                <styles.DeleteButton
                  onClick={(e) => {
                    deleteComment(index);
                  }}
                  data-testid={'deleteButton'}
                />
              </styles.LatestSearch>
            ))
          )
        ) : null}
      </styles.LatestSearchListBlock>
    </styles.MainBlock>
  );
};

export default SearchBarPresentational;
