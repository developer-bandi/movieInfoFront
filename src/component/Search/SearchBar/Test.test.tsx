import { renderHook } from '@testing-library/react-hooks';
import { act } from 'react-dom/test-utils';
import { Provider } from 'react-redux';
import useSearchBar from './Hook';
import configureMockStore from 'redux-mock-store';
import { fireEvent, render, screen } from '@testing-library/react';
import SearchBar from './Presentational';

describe('SearchBar Presentational', () => {
  it('이벤트 핸들러 동작 체크', () => {
    const searchMoviesMock = jest.fn();
    const deleteCommentMock = jest.fn();
    const settingvalueMock = jest.fn();
    const setFocusMock = jest.fn();
    render(
      <SearchBar
        latest={['test']}
        searchMovies={searchMoviesMock}
        deleteComment={deleteCommentMock}
        settingvalue={settingvalueMock}
        value={'test'}
        focus={true}
        setFocus={setFocusMock}
      />
    );
    fireEvent.click(screen.getByTestId('searchInput'), {
      target: { value: 'test' },
    });
    fireEvent.keyPress(screen.getByTestId('searchInput'), {
      key: 'Enter',
      code: 'Enter',
      charCode: 13,
    });
    fireEvent.change(screen.getByTestId('searchInput'), {
      target: { value: 'new value' },
    });
    fireEvent.click(screen.getByTestId('searchButton'));
    fireEvent.click(screen.getByTestId('data'));
    fireEvent.click(screen.getByTestId('deleteButton'));
    expect(searchMoviesMock).toBeCalledTimes(3);
    expect(deleteCommentMock).toBeCalledTimes(1);
    expect(setFocusMock).toBeCalledTimes(1);
    expect(settingvalueMock).toBeCalledTimes(1);
  });
  it('최근 검색어 비활성화', () => {
    const searchMoviesMock = jest.fn();
    const deleteCommentMock = jest.fn();
    const settingvalueMock = jest.fn();
    const setFocusMock = jest.fn();
    const utils = render(
      <SearchBar
        latest={['test']}
        searchMovies={searchMoviesMock}
        deleteComment={deleteCommentMock}
        settingvalue={settingvalueMock}
        value={'test'}
        focus={false}
        setFocus={setFocusMock}
      />
    );
    expect(utils.container).toMatchSnapshot();
  });
  describe('최근 검색어 활성화', () => {
    it('최근 검색어 없음', () => {
      const searchMoviesMock = jest.fn();
      const deleteCommentMock = jest.fn();
      const settingvalueMock = jest.fn();
      const setFocusMock = jest.fn();
      const utils = render(
        <SearchBar
          latest={[]}
          searchMovies={searchMoviesMock}
          deleteComment={deleteCommentMock}
          settingvalue={settingvalueMock}
          value={'test'}
          focus={true}
          setFocus={setFocusMock}
        />
      );
      expect(utils.container).toMatchSnapshot();
      screen.getByText('최근 검색 내용이 없습니다');
    });
    it('최근 검색어 있음', () => {
      const searchMoviesMock = jest.fn();
      const deleteCommentMock = jest.fn();
      const settingvalueMock = jest.fn();
      const setFocusMock = jest.fn();
      const utils = render(
        <SearchBar
          latest={['test']}
          searchMovies={searchMoviesMock}
          deleteComment={deleteCommentMock}
          settingvalue={settingvalueMock}
          value={'test'}
          focus={true}
          setFocus={setFocusMock}
        />
      );
      expect(utils.container).toMatchSnapshot();
      screen.getByText('test');
    });
  });
});

describe('SearchBar Hook', () => {
  const mockStore = configureMockStore();
  const store = mockStore({});
  const wrapper = ({ children }: { children: React.FC }) => (
    <Provider store={store}>{children}</Provider>
  );

  it('localstorage 데이터 가져오는 useEffect 체크', () => {
    localStorage.setItem('latest', JSON.stringify(['test']));
    const { result } = renderHook(() => useSearchBar(), {
      wrapper,
    });
    expect(result.current.latest).toStrictEqual(['test']);
  });
  describe('searchMoviesg함수 테스트', () => {
    it('검색 버튼 클릭시 검색 기능 작동여부', () => {
      localStorage.setItem('latest', JSON.stringify(['test1', 'test2']));
      const { result } = renderHook(() => useSearchBar(), {
        wrapper,
      });
      act(() => {
        result.current.setvalue('test3');
      });
      act(() => {
        result.current.searchMovies({
          type: 'click',
          target: { innerText: 'test3', tagName: 'BUTTON' },
        });
      });
      expect(store.getActions()[0]).toStrictEqual({
        type: 'movieSearch/searchMovie',
        payload: { keyword: 'test3', page: 1 },
      });
      expect(
        JSON.parse(localStorage.getItem('latest') as string)
      ).toStrictEqual(['test1', 'test2', 'test3']);
    });

    it('엔터 버튼 클릭시 검색기능 작동 여부', () => {
      localStorage.setItem(
        'latest',
        JSON.stringify(['test1', 'test2', 'test3'])
      );
      const { result } = renderHook(() => useSearchBar(), {
        wrapper,
      });
      act(() => {
        result.current.setvalue('test4');
      });
      act(() => {
        result.current.searchMovies({
          key: 'Enter',
          type: 'keypress',
          target: { innerText: 'test4', tagName: 'BUTTON' },
        });
      });
      expect(store.getActions()[1]).toStrictEqual({
        type: 'movieSearch/searchMovie',
        payload: { keyword: 'test4', page: 1 },
      });
      expect(
        JSON.parse(window.localStorage.getItem('latest') as string)
      ).toStrictEqual(['test1', 'test2', 'test3', 'test4']);
    });

    it('최근 검색어 클릭시 검색기능 작동여부', () => {
      window.localStorage.setItem(
        'latest',
        JSON.stringify(['test1', 'test2', 'test3', 'test4'])
      );
      const { result } = renderHook(() => useSearchBar(), {
        wrapper,
      });
      act(() => {
        result.current.searchMovies({
          type: 'click',
          target: { innerText: 'test5', tagName: 'DIV' },
        });
      });
      expect(store.getActions()[2]).toStrictEqual({
        type: 'movieSearch/searchMovie',
        payload: { keyword: 'test5', page: 1 },
      });
      expect(
        JSON.parse(window.localStorage.getItem('latest') as string)
      ).toStrictEqual(['test1', 'test2', 'test3', 'test4', 'test5']);
    });
    it('최근 검색어가 5개일경우 가장 이전 검색어를 지우고 최신 검색 반영을 반영하는지', () => {
      localStorage.setItem(
        'latest',
        JSON.stringify(['test1', 'test2', 'test3', 'test4', 'test5'])
      );
      const { result } = renderHook(() => useSearchBar(), {
        wrapper,
      });
      act(() => {
        result.current.setvalue('new Test');
      });
      act(() => {
        result.current.searchMovies({
          type: 'click',
          target: { innerText: 'new Test', tagName: 'DIV' },
        });
      });
      expect(store.getActions()[3]).toStrictEqual({
        type: 'movieSearch/searchMovie',
        payload: { keyword: 'new Test', page: 1 },
      });
      expect(
        JSON.parse(window.localStorage.getItem('latest') as string)
      ).toStrictEqual(['test2', 'test3', 'test4', 'test5', 'new Test']);
    });
  });
});
