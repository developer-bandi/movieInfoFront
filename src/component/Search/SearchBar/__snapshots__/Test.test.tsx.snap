// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SearchBar Presentational 최근 검색어 비활성화 1`] = `
<div>
  <div
    class="sc-bdvvtL dEqaoN"
  >
    <div
      class="sc-gsDKAQ chIZOY"
    >
      <input
        class="sc-dkPtRN OPwlS"
        data-testid="searchInput"
        type="text"
        value="test"
      />
      <svg
        class="sc-hKwDye ceUqMh"
        data-testid="searchButton"
        fill="currentColor"
        height="1em"
        stroke="currentColor"
        stroke-width="0"
        viewBox="0 0 1024 1024"
        width="1em"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M909.6 854.5L649.9 594.8C690.2 542.7 712 479 712 412c0-80.2-31.3-155.4-87.9-212.1-56.6-56.7-132-87.9-212.1-87.9s-155.5 31.3-212.1 87.9C143.2 256.5 112 331.8 112 412c0 80.1 31.3 155.5 87.9 212.1C256.5 680.8 331.8 712 412 712c67 0 130.6-21.8 182.7-62l259.7 259.6a8.2 8.2 0 0 0 11.6 0l43.6-43.5a8.2 8.2 0 0 0 0-11.6zM570.4 570.4C528 612.7 471.8 636 412 636s-116-23.3-158.4-65.6C211.3 528 188 471.8 188 412s23.3-116.1 65.6-158.4C296 211.3 352.2 188 412 188s116.1 23.2 158.4 65.6S636 352.2 636 412s-23.3 116.1-65.6 158.4z"
        />
      </svg>
    </div>
    <div
      class="sc-eCImPb jinBsB"
    />
  </div>
</div>
`;

exports[`SearchBar Presentational 최근 검색어 활성화 최근 검색어 없음 1`] = `
<div>
  <div
    class="sc-bdvvtL dEqaoN"
  >
    <div
      class="sc-gsDKAQ chIZOY"
    >
      <input
        class="sc-dkPtRN OPwlS"
        data-testid="searchInput"
        type="text"
        value="test"
      />
      <svg
        class="sc-hKwDye ceUqMh"
        data-testid="searchButton"
        fill="currentColor"
        height="1em"
        stroke="currentColor"
        stroke-width="0"
        viewBox="0 0 1024 1024"
        width="1em"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M909.6 854.5L649.9 594.8C690.2 542.7 712 479 712 412c0-80.2-31.3-155.4-87.9-212.1-56.6-56.7-132-87.9-212.1-87.9s-155.5 31.3-212.1 87.9C143.2 256.5 112 331.8 112 412c0 80.1 31.3 155.5 87.9 212.1C256.5 680.8 331.8 712 412 712c67 0 130.6-21.8 182.7-62l259.7 259.6a8.2 8.2 0 0 0 11.6 0l43.6-43.5a8.2 8.2 0 0 0 0-11.6zM570.4 570.4C528 612.7 471.8 636 412 636s-116-23.3-158.4-65.6C211.3 528 188 471.8 188 412s23.3-116.1 65.6-158.4C296 211.3 352.2 188 412 188s116.1 23.2 158.4 65.6S636 352.2 636 412s-23.3 116.1-65.6 158.4z"
        />
      </svg>
    </div>
    <div
      class="sc-eCImPb jinBsB"
    >
      <div
        class="sc-jRQBWg bYIsxb"
      >
        최근 검색 내용이 없습니다
      </div>
    </div>
  </div>
</div>
`;

exports[`SearchBar Presentational 최근 검색어 활성화 최근 검색어 있음 1`] = `
<div>
  <div
    class="sc-bdvvtL dEqaoN"
  >
    <div
      class="sc-gsDKAQ chIZOY"
    >
      <input
        class="sc-dkPtRN OPwlS"
        data-testid="searchInput"
        type="text"
        value="test"
      />
      <svg
        class="sc-hKwDye ceUqMh"
        data-testid="searchButton"
        fill="currentColor"
        height="1em"
        stroke="currentColor"
        stroke-width="0"
        viewBox="0 0 1024 1024"
        width="1em"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M909.6 854.5L649.9 594.8C690.2 542.7 712 479 712 412c0-80.2-31.3-155.4-87.9-212.1-56.6-56.7-132-87.9-212.1-87.9s-155.5 31.3-212.1 87.9C143.2 256.5 112 331.8 112 412c0 80.1 31.3 155.5 87.9 212.1C256.5 680.8 331.8 712 412 712c67 0 130.6-21.8 182.7-62l259.7 259.6a8.2 8.2 0 0 0 11.6 0l43.6-43.5a8.2 8.2 0 0 0 0-11.6zM570.4 570.4C528 612.7 471.8 636 412 636s-116-23.3-158.4-65.6C211.3 528 188 471.8 188 412s23.3-116.1 65.6-158.4C296 211.3 352.2 188 412 188s116.1 23.2 158.4 65.6S636 352.2 636 412s-23.3 116.1-65.6 158.4z"
        />
      </svg>
    </div>
    <div
      class="sc-eCImPb jinBsB"
    >
      <div
        class="sc-gKclnd hsyfwU"
      >
        <div
          data-testid="data"
        >
          test
        </div>
        <svg
          class="sc-iCfMLu kQMrYh"
          data-testid="deleteButton"
          fill="currentColor"
          height="1em"
          stroke="currentColor"
          stroke-width="0"
          viewBox="0 0 16 16"
          width="1em"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M1.293 1.293a1 1 0 0 1 1.414 0L8 6.586l5.293-5.293a1 1 0 1 1 1.414 1.414L9.414 8l5.293 5.293a1 1 0 0 1-1.414 1.414L8 9.414l-5.293 5.293a1 1 0 0 1-1.414-1.414L6.586 8 1.293 2.707a1 1 0 0 1 0-1.414z"
          />
        </svg>
      </div>
    </div>
  </div>
</div>
`;
